From e97c30444cc1c73d32e418161c9783acc89c8bca Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Thu, 27 Mar 2025 16:24:16 +0100
Subject: [PATCH 1644/1695] plat-stm32mp1: clean WKUP IRQ management

Clean the PWR WKUP management, as it is described in reference manual:
- clear flag before enable to avoid false interrupt
- don't disable PWR interruption for mask operations because
  mask the parent interrupt is enough for interrupted framework
  and this patch avoids to lost a PWR interruption between
  call of mask and unmask function, when interruption are disabled.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 89b0d6cdc853bc62e9ec5c11d262c278fde63ee1 plat-stm32mp1: clean WKUP IRQ management]
Change-Id: I1829f132d021a8931d71787f2962dab37a8f6eec
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/447259
Reviewed-by: Antonio Maria BORNEO <antonio.borneo@foss.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
index 6c97b9d1b..bf800c784 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr_irq.c
@@ -150,6 +150,8 @@ static void stm32mp1_pwr_itr_enable_nolock(size_t it)
 
 	VERBOSE_PWR("Pwr irq enable");
 
+	/* Clear flag before enable to avoid false interrupt */
+	io_setbits32(priv->base + WKUPCR, BIT(it));
 	io_setbits32(priv->base + MPUWKUPENR, BIT(it));
 }
 
@@ -239,7 +241,6 @@ static void stm32mp1_pwr_op_mask(struct itr_chip *chip __unused, size_t it)
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
 	priv->itr_mask_bitmask |= BIT(it);
-	stm32mp1_pwr_itr_disable_nolock(it);
 	interrupt_mask(priv->parent_hdl->chip, priv->parent_hdl->it);
 	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
 }
@@ -252,7 +253,6 @@ static void stm32mp1_pwr_op_unmask(struct itr_chip *chip __unused, size_t it)
 
 	exceptions = cpu_spin_lock_xsave(&priv->spinlock);
 	priv->itr_mask_bitmask &= ~BIT(it);
-	stm32mp1_pwr_itr_enable_nolock(it);
 	if (!priv->itr_mask_bitmask)
 		interrupt_unmask(priv->parent_hdl->chip, priv->parent_hdl->it);
 	cpu_spin_unlock_xrestore(&priv->spinlock, exceptions);
-- 
2.47.2

